File: Android/BarkCloud.Android/app/src/main/java/com/barkfluff/BarkCloud/data/users/UserRepository.kt
package com.barkfluff.BarkCloud.data.users

import android.net.Uri
import barkcloud.files.FilesApiOuterClass.UploadFileType
import barkcloud.users.UsersApiOuterClass.ChangeBioRequest
import barkcloud.users.UsersApiOuterClass.ChangeNameRequest
import barkcloud.users.UsersApiOuterClass.ChangeUsernameRequest
import barkcloud.users.UsersApiOuterClass.CheckExistUsernameRequest
import barkcloud.users.UsersApiOuterClass.DeleteAccountRequest
import barkcloud.users.UsersApiOuterClass.DeleteDeviceRequest
import barkcloud.users.UsersApiOuterClass.Device
import barkcloud.users.UsersApiOuterClass.GetCurrentDeviceRequest
import barkcloud.users.UsersApiOuterClass.GetDevicesRequest
import barkcloud.users.UsersApiOuterClass.GetPrivacySettingsRequest
import barkcloud.users.UsersApiOuterClass.GetUserRequest
import barkcloud.users.UsersApiOuterClass.PrivacySettings
import barkcloud.users.UsersApiOuterClass.RenameDeviceRequest
import barkcloud.users.UsersApiOuterClass.SetProfilePictureRequest
import barkcloud.users.UsersApiOuterClass.UpdatePrivacySettingsRequest
import barkcloud.users.UsersApiOuterClass.User
import com.barkfluff.BarkCloud.grpc.GrpcManager
import com.barkfluff.BarkCloud.net.FileTransferService

/**
 * Доступ к сервису Users (профиль, приватность, устройства, аккаунт) + установка
 * аватара (через [FileTransferService]). Методы возвращают proto-типы — UI читает поля
 * напрямую (как в iOS).
 */
class UserRepository(
    private val grpc: GrpcManager,
    private val transfer: FileTransferService,
) {

    // MARK: Профиль

    /** Профиль пользователя. `userId = 0` — свой профиль. */
    suspend fun getUser(userId: Long = 0): User =
        grpc.usersStub().getUser(GetUserRequest.newBuilder().setUserId(userId).build()).user

    suspend fun changeName(firstName: String, lastName: String) {
        grpc.usersStub().changeName(
            ChangeNameRequest.newBuilder().setFirstName(firstName).setLastName(lastName).build()
        )
    }

    suspend fun usernameExists(username: String): Boolean =
        grpc.usersStub().checkExistUsername(
            CheckExistUsernameRequest.newBuilder().setUsername(username).build()
        ).exist

    suspend fun changeUsername(username: String) {
        grpc.usersStub().changeUsername(
            ChangeUsernameRequest.newBuilder().setUsername(username).build()
        )
    }

    suspend fun changeBio(bio: String) {
        grpc.usersStub().changeBio(ChangeBioRequest.newBuilder().setBio(bio).build())
    }

    // MARK: Приватность

    suspend fun getPrivacySettings(): PrivacySettings =
        grpc.usersStub().getPrivacySettings(GetPrivacySettingsRequest.getDefaultInstance()).settings

    suspend fun updatePrivacySettings(settings: PrivacySettings): PrivacySettings =
        grpc.usersStub().updatePrivacySettings(
            UpdatePrivacySettingsRequest.newBuilder().setSettings(settings).build()
        ).settings

    // MARK: Устройства

    suspend fun getDevices(): List<Device> =
        grpc.usersStub().getDevices(GetDevicesRequest.getDefaultInstance()).devicesList

    suspend fun getCurrentDevice(): Device? {
        val resp = grpc.usersStub().getCurrentDevice(GetCurrentDeviceRequest.getDefaultInstance())
        return if (resp.hasDevice()) resp.device else null
    }

    suspend fun renameDevice(deviceId: String, customName: String) {
        grpc.usersStub().renameDevice(
            RenameDeviceRequest.newBuilder().setDeviceId(deviceId).setCustomName(customName).build()
        )
    }

    suspend fun deleteDevice(deviceId: String) {
        grpc.usersStub().deleteDevice(DeleteDeviceRequest.newBuilder().setDeviceId(deviceId).build())
    }

    // MARK: Аккаунт

    suspend fun deleteAccount() {
        grpc.usersStub().deleteAccount(DeleteAccountRequest.getDefaultInstance())
    }

    // MARK: Аватар (двухшаговый флоу)

    /** Загрузить картинку как `USER_AVATAR` и установить аватаром. */
    suspend fun setAvatar(uri: Uri, fileName: String) {
        val target = transfer.getUploadUrl(UploadFileType.USER_AVATAR)
        val fileId = transfer.upload(uri, fileName, target.url)
        grpc.usersStub().setProfilePicture(
            SetProfilePictureRequest.newBuilder().setFileId(fileId).build()
        )
    }

    /** Удалить аватар (`SetProfilePicture("")`). */
    suspend fun removeAvatar() {
        grpc.usersStub().setProfilePicture(
            SetProfilePictureRequest.newBuilder().setFileId("").build()
        )
    }
}

File: Ios/BarkCloud/BarkCloud/Data/Users/UserRepository.swift
import Foundation
import GRPCCore

/// Доступ к сервису Users (профиль, приватность, устройства, аккаунт) + установка
/// аватара (через `FileTransferService`). Методы пробрасывают `RPCError` —
/// доменные ошибки маппит UI через `domainErrorMessage(_:)`.
final class UserRepository: Sendable {
    private let grpc: GrpcManager
    private let transfer: FileTransferService

    init(grpc: GrpcManager, transfer: FileTransferService) {
        self.grpc = grpc
        self.transfer = transfer
    }

    // MARK: - Профиль

    /// Профиль пользователя. `userID = 0` — свой профиль.
    func getUser(userID: Int64 = 0) async throws -> Barkcloud_Users_User {
        let stub = try await grpc.usersStub()
        var req = Barkcloud_Users_GetUserRequest()
        req.userID = userID
        return try await stub.getUser(req).user
    }

    func changeName(firstName: String, lastName: String) async throws {
        let stub = try await grpc.usersStub()
        var req = Barkcloud_Users_ChangeNameRequest()
        req.firstName = firstName
        req.lastName = lastName
        _ = try await stub.changeName(req)
    }

    func usernameExists(_ username: String) async throws -> Bool {
        let stub = try await grpc.usersStub()
        var req = Barkcloud_Users_CheckExistUsernameRequest()
        req.username = username
        return try await stub.checkExistUsername(req).exist
    }

    func changeUsername(_ username: String) async throws {
        let stub = try await grpc.usersStub()
        var req = Barkcloud_Users_ChangeUsernameRequest()
        req.username = username
        _ = try await stub.changeUsername(req)
    }

    func changeBio(_ bio: String) async throws {
        let stub = try await grpc.usersStub()
        var req = Barkcloud_Users_ChangeBioRequest()
        req.bio = bio
        _ = try await stub.changeBio(req)
    }

    // MARK: - Поиск

    /// Поиск пользователей по юзернейму/имени. Сервер требует минимум 2 символа,
    /// поэтому короткие запросы заворачиваем тут же без сетевого вызова.
    /// `limit` 1..50 (default 20). Возвращает только тех, у кого
    /// `PrivacySettings.searchableByUsername == true`.
    func searchUsers(query: String, limit: Int = 20) async throws -> [CloudUser] {
        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed.count >= 2 else { return [] }
        let stub = try await grpc.usersStub()
        var req = Barkcloud_Users_SearchUsersRequest()
        req.query = trimmed
        req.limit = Int32(max(1, min(50, limit)))
        let resp = try await stub.searchUsers(req)
        return resp.users.map(CloudUser.init)
    }

    // MARK: - Приватность

    func getPrivacySettings() async throws -> Barkcloud_Users_PrivacySettings {
        let stub = try await grpc.usersStub()
        return try await stub.getPrivacySettings(Barkcloud_Users_GetPrivacySettingsRequest()).settings
    }

    @discardableResult
    func updatePrivacySettings(_ settings: Barkcloud_Users_PrivacySettings) async throws -> Barkcloud_Users_PrivacySettings {
        let stub = try await grpc.usersStub()
        var req = Barkcloud_Users_UpdatePrivacySettingsRequest()
        req.settings = settings
        return try await stub.updatePrivacySettings(req).settings
    }

    // MARK: - Устройства

    func getDevices() async throws -> [Barkcloud_Users_Device] {
        let stub = try await grpc.usersStub()
        return try await stub.getDevices(Barkcloud_Users_GetDevicesRequest()).devices
    }

    func getCurrentDevice() async throws -> Barkcloud_Users_Device? {
        let stub = try await grpc.usersStub()
        let resp = try await stub.getCurrentDevice(Barkcloud_Users_GetCurrentDeviceRequest())
        return resp.hasDevice ? resp.device : nil
    }

    func renameDevice(deviceID: String, customName: String) async throws {
        let stub = try await grpc.usersStub()
        var req = Barkcloud_Users_RenameDeviceRequest()
        req.deviceID = deviceID
        req.customName = customName
        _ = try await stub.renameDevice(req)
    }

    func deleteDevice(deviceID: String) async throws {
        let stub = try await grpc.usersStub()
        var req = Barkcloud_Users_DeleteDeviceRequest()
        req.deviceID = deviceID
        _ = try await stub.deleteDevice(req)
    }

    // MARK: - Аккаунт

    func deleteAccount() async throws {
        let stub = try await grpc.usersStub()
        _ = try await stub.deleteAccount(Barkcloud_Users_DeleteAccountRequest())
    }

    // MARK: - Аватар (двухшаговый флоу)

    /// Загрузить картинку как `USER_AVATAR` и установить аватаром.
    func setAvatar(imageData: Data, fileName: String) async throws {
        let upload = try await transfer.getUploadURL(type: .userAvatar)
        let fileID = try await transfer.upload(data: imageData, fileName: fileName, to: upload.url)
        let stub = try await grpc.usersStub()
        var req = Barkcloud_Users_SetProfilePictureRequest()
        req.fileID = fileID
        _ = try await stub.setProfilePicture(req)
    }

    /// Удалить аватар (`SetProfilePicture("")`).
    func removeAvatar() async throws {
        let stub = try await grpc.usersStub()
        var req = Barkcloud_Users_SetProfilePictureRequest()
        req.fileID = ""
        _ = try await stub.setProfilePicture(req)
    }
}
